lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Delegation.hs

```diff
@@ -0,0 +1,464 @@
+-- @
+data DelegationState k = DelegationState
+    { -- | The account public key from which the stake keys should be derived.
+      rewardAccountKey :: k 'AccountK XPub
+    , state :: State
+    } deriving (Generic)
+
+
+
+-- | The internal state, without the account key.
+--
+-- TODO: Perhaps we should model this as
+-- @S = S1 * S2, where S1 = Bool, S2 = ix * UTxOPointer@ - having two
+-- "concurrent" states tracking stake keys, where the first one is identical to
+-- legacy single-stake key wallets.
+--
+-- Maybe that would help simplify `applyTx` and `setPortfolioOf`...
+data State
+    -- | No active stake keys. The initial state of a new wallet.
+    = Zero
+
+    -- | The first stake-key (index 0) is registered and either delegating or
+    -- about to be delegating.
+    | One
+
+    -- | There is more than one active stake keys. Can only be reached using
+    -- wallets with support for multiple stake keys.
+    | More
+        !(Index 'Soft 'AddressK)
+          -- nextKeyIx - the ix of the next unused key
+        PointerUTxO
+          -- ^ pointer utxo that need to be spent when changing state.
+        !Bool
+          -- ^ Is key 0 still registered? For compatibility with
+          -- single-stake-key wallets, we need to track this.
+          --
+          -- See the implementaiton of @applyTx@ for how it is used.
+    deriving (Eq, Show, Generic)
+
+instance NFData State
+
+instance (NFData (k 'AccountK XPub), NFData (k 'AddressK XPub))
+    => NFData (DelegationState k)
+
+deriving instance
+    ( Show (k 'AccountK XPub)
+    , Show (k 'AddressK XPub)
+    ) => Show (DelegationState k)
+
+deriving instance
+    ( Eq (k 'AccountK XPub)
+    , Eq (k 'AddressK XPub)
+    ) => Eq (DelegationState k)
+
+pointer :: DelegationState k -> Maybe PointerUTxO
+pointer (DelegationState _ (More _ p _)) = Just p
+pointer _ = Nothing
+
+keyAtIx
+    :: (SoftDerivation k)
+    => DelegationState k
+    -> Index 'Soft 'AddressK
+    -> k 'AddressK XPub
+keyAtIx s = deriveAddressPublicKey (rewardAccountKey s) MutableAccount
+
+nextKeyIx
+    :: DelegationState k
+    -> Index 'Soft 'AddressK
+nextKeyIx s = case state s of
+    Zero -> minBound
+    One -> succ minBound
+    More ix _ _ -> ix
+
+lastActiveIx
+    :: DelegationState k
+    -> Maybe (Index 'Soft 'AddressK)
+lastActiveIx s
+    | nextKeyIx s == minBound = Nothing
+    | otherwise               = Just $ pred $ nextKeyIx s
+
+data PointerUTxO = PointerUTxO { pTxIn :: TxIn, pCoin :: Coin }
+    deriving (Generic, Eq, Show)
+    deriving anyclass NFData
+
+-- | Returns the index corresponding to the payment key the `PointerUTxO`
+-- should be locked with for a portfolio of a given size @n@.
+--
+-- In our current implementation we require the `PointerUTxO` to be created in
+-- the @[0] -> [0,1] transition@, i.e. @nextKeyIx 1 -> nextKeyIx 2@.
+pointerIx
+    :: Int
+    -> Maybe (Index 'Soft 'AddressK)
+pointerIx 0 = Nothing
+pointerIx 1 = Nothing
+pointerIx n = Just $ toEnum n
```

Comment:
Now we have
`[0,1], 1 ada in held by key 2`
because I previously allowed
`[], 1 ada in held by key 0`

Now the latter state is not possible. So we _could_ have the pointer correspond to to the `lastActive` rather than the `nextUnused` ix. I.e.
`[0,1], 1 ada in held by key 1`

Reply:
I am a little bit confused by this function at the moment, maybe it will get clearer in the rest of the review. However, I would expect the pointer to be created during the transition [0] -> [0,1] (as you state in comment) and that, `pointerIx n = Just n`; It feels more intuitive to me for the pointer to track the _highest active key_.

Reply:
> It feels more intuitive to me for the pointer to track the highest active key

Yeah, I was starting to feel the same. Might try that. Here it would be `pointerIx n = Just $ pred $ toEnum n` though, as `n` is the number of stake keys you want.

That said, there should be some _small_ fee-tradeoffs from this choice (if I understand correctly)

With the pointer tracking `nextKeyIx`:
- Adding stake keys: we need to spend the pointer at `nextKeyIx`, _and_ delegate with `nextKeyIx`, so we save a witness.
- Removing stake keys: no savings

With the pointer tracking `lastActiveIx`:
- Adding stake keys: no savings
- Removing stake keys: we need to spend the pointer at `lastActiveIx`, _and_ deregister with `lastActiveIx`, so we save a witness.

But I guess the difference is so small it's not even worth thinking about with 44 lovelace/byte.

Reply:
For the record, a argument in the other direction I didn't think of until recently:

Assuming
- re-delegation of existing stake-keys requires spending and re-creating the pointer
- the pointer uses `lastActiveIx`

Then we save a witness when changing the delegation at `lastActiveIx`.